client/zone/js-modules/map-color-finder.js

```diff
@@ -0,0 +1,198 @@
+    createPixelToSegmentMapping(preparedLayers) {
+        var pixelData = this.create2DArray(
+            preparedLayers.boundaries.maxX + 1,
+            preparedLayers.boundaries.maxY + 1
+        );
+        var numberOfSegments = 0;
+        preparedLayers.segments.forEach(seg => {
+            numberOfSegments += 1;
+            seg.pixels.forEach(p => {
+                pixelData[p.x][p.y] = seg.segmentIndex;
+            });
+        });
+        return {
+            map: pixelData,
+            numberOfSegments: numberOfSegments,
+            boundaries: preparedLayers.boundaries
+        };
+    }
+
+    buildGraph(mapData) {
+        var vertices = this.makeArray(mapData.numberOfSegments).map(i => new MapAreaVertex(i));
+        var graph = new MapAreaGraph(vertices);
+        this.traverseMap(mapData.boundaries, mapData.map, (x, y, currentSegmentId, pixelData) => {
+            var newSegmentId = pixelData[x][y];
+            graph.connectVertices(currentSegmentId, newSegmentId);
+            return newSegmentId != undefined ? newSegmentId : currentSegmentId;
+        });
+        return graph;
+    }
+
+    traverseMap(boundaries, pixelData, func) {
+        // row-first traversal        
+        for (let y = boundaries.minY; y <= boundaries.maxY; y++) {
+            var currentSegmentId = undefined;
+            for (let x = boundaries.minX; x <= boundaries.maxX; x++) {
+                currentSegmentId = func(x, y, currentSegmentId, pixelData);
+            }
+        }
+        // column-first traversal
+        for (let x = boundaries.minX; x <= boundaries.maxX; x++) {
+            var currentSegmentId = undefined;
+            for (let y = boundaries.minY; y <= boundaries.maxY; y++) {
+                currentSegmentId = func(x, y, currentSegmentId, pixelData);
+            }
+        }
+    }
+
+    /**
+     * Credit for this function goes to the authors of this StackOverflow answer: https://stackoverflow.com/a/966938
+     */
+    create2DArray(length) {
+        var arr = new Array(length || 0),
+            i = length;
+        if (arguments.length > 1) {
+            var args = Array.prototype.slice.call(arguments, 1);
+            while (i--) arr[length - 1 - i] = this.create2DArray.apply(this, args);
+        }
+        return arr;
+    }
+
+    makeArray(n) {
+        return Array.apply(null, { length: n }).map(Number.call, Number);
+    }
+}
+
+class MapAreaVertex {
+    constructor(id) {
+        this.id = id;
+        this.adjacentVertexIds = [];
+        this.color = undefined;
+    }
+
+    appendVertex(vertexId) {
+        if (vertexId != undefined && !this.adjacentVertexIds.includes(vertexId)) {
+            this.adjacentVertexIds.push(vertexId);
+        }
+    }
+}
+
+class MapAreaGraph {
+    constructor(vertices) {
+        this.vertices = vertices;
+    }
+
+    connectVertices(id1, id2) {
+        if (id1 != undefined && id2 != undefined && id1 != id2) {
+            this.getById(id1)?.appendVertex(id2);
```

Comment:
https://caniuse.com/#feat=mdn-javascript_operators_optional_chaining

We probably can't use this syntax yet :(

Reply:
It's you and the linter against me, so I'll have to comply.